src/ode/man/man3/def_build.3
.\" 
.\" @OSF_FREE_COPYRIGHT@
.\" COPYRIGHT NOTICE
.\" Copyright (c) 1992, 1991, 1990  
.\" Open Software Foundation, Inc. 
.\"  
.\" Permission is hereby granted to use, copy, modify and freely distribute 
.\" the software in this file and its documentation for any purpose without 
.\" fee, provided that the above copyright notice appears in all copies and 
.\" that both the copyright notice and this permission notice appear in 
.\" supporting documentation.  Further, provided that the name of Open 
.\" Software Foundation, Inc. ("OSF") not be used in advertising or 
.\" publicity pertaining to distribution of the software without prior 
.\" written permission from OSF.  OSF makes no representations about the 
.\" suitability of this software for any purpose.  It is provided "as is" 
.\" without express or implied warranty. 
.\"
.\"
.\" HISTORY
.\" $Log: def_build.3,v $
.\" Revision 1.2.2.2  1992/12/03  17:25:00  damon
.\" 	ODE 2.2 CR 183. Added CMU notice
.\" 	[1992/12/03  17:10:33  damon]
.\"
.\" Revision 1.2  1991/12/05  21:16:03  devrcs
.\" 	Man page for routines: default_build, get_rc_value,
.\" 	buid_base_dir, and rm_newline.
.\" 	[91/01/30  16:47:38  randyb]
.\" 
.\" $EndLog$
.\" 
.\" $EndLog$
.\" 
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.TH def_build 3 1/28/91
.SH NAME
def_build \- holds routines which access rc build information
.SH SYNOPSIS
.IP "\fBint default_build ( buildbase, buildname, buildset, sb_rcfile )\fR"
.sp 0
.B "char ** buildbase;"
.sp 0
.B "char ** buildname;"
.sp 0
.B "char ** buildset;"
.sp 0
.B "char * sb_rcfile;"
.IP "\fBint get_rc_value ( keyword, value, contents, msgs )\fR"
.sp 0
.B "char * keyword;"
.sp 0
.B "char ** output;"
.sp 0
.B "struct rcfile contents;"
.sp 0
.B "BOOLEAN msgs;"
.IP "\fBchar * build_base_dir ( keyword, contents, errors, default )\fR"
.sp 0
.B "char * keyword;"
.sp 0
.B "struct rcfile contents;"
.sp 0
.B "BOOLEAN errors;"
.sp 0
.B "BOOLEAN default"
.IP "\fBvoid rm_newline ( char* string )"
.sp 0
.B "char * string;"
.SH DESCRIPTION
.SS \fBdefault_build\fR
\fBdefault_build\fR fills in the three arguments; \fBbuildbase, buildname,\fR
and \fBbuildset\fR, with the values from the sandbox rc file \fBsb_rcfile\fR.
These values correspond to the three rc file variables:
\fBbuild_base\fR, \fBdefault_build\fR, and \fBdefault_set\fR.
If an argument is set to NUL, '\0', no attempt is made to fill it.
If the argument is not NUL, \fBdefault_build\fR will create space for
for the value and set the char pointer to it.
.P
The build base is the directory the project's builds are under.
The default build is the build submissions are made to, and the default
set, the set used for submissions.

.SS \fBget_rc_value\fR
\fBget_rc_value\fR reads through the contents of \fBcontents\fR
and returns the value of the \fBkeyword\fR.
It then creates space for this value and points \fB*output\fR at
this new space.
.P
The structure \fBcontents\fR must previously have been initialized
with a call to \fBparse_rc_file\fR.
.P
If \fBmsgs\fR is set to TRUE and there is a failure in this process,
the routine prints messages about why it failed.

.SS \fBbuild_base_dir\fR
\fBbuild_base_dir\fR looks for the base directory of the build specified
by \fBkeyword\fR.
If it finds the directory, it returns a pointer to the string with directory
path, else it returns NUL.
\fBbuild_base_dir\fR uses the value of the sandbox rc variable \fBbuild_list\fR
to determine what file to read for the information it is searching for,
or, if \fBbuild_list\fR is not set, it looks in the file
\fB/project/projects/build_list\fR.
.P
If \fBerrors\fR is TRUE, this routine prints error messages when it fails,
for any reason, to find the base directory for the \fBkeyword\fR.
If \fBdefault\fR is TRUE, this routine will use the default build list
and not search for the one in the backing tree.
.P
\fBcontents\fR is an initialized rcfile structure which \fBbuild_base_dir\fR
will use to get the name of the build list file.

.SS \fBrm_newline\fR
This function removes the last occurrence of newline, '\n', from a string
and replaces it with NUL, '\0'.
.SH FILES
.IP \fBlibsb.a\fR
library containing these routines
.IP \fB/project/projects/build_list\fR
the default, site file to hold the list of builds, their configuration dates,
and base directories.
.SH EXIT VALUES
\fBdefault_build\fR returns '0' upon successful completion, '-1' otherwise.
.br
\fBget_rc_value\fR returns '0' upon successful completion, '-1' otherwise.
.br
\fBbuild_base_dir\fR returns a pointer to the string containing the
base directory upon successful completion; it returns NULL if for any
reason it cannot get this information.
.br
\fBrm_newline\fR does not return a value.


src/ode/lib/libode/def_build.c
/*
 * @OSF_FREE_COPYRIGHT@
 * COPYRIGHT NOTICE
 * Copyright (c) 1992, 1991, 1990  
 * Open Software Foundation, Inc. 
 *  
 * Permission is hereby granted to use, copy, modify and freely distribute 
 * the software in this file and its documentation for any purpose without 
 * fee, provided that the above copyright notice appears in all copies and 
 * that both the copyright notice and this permission notice appear in 
 * supporting documentation.  Further, provided that the name of Open 
 * Software Foundation, Inc. ("OSF") not be used in advertising or 
 * publicity pertaining to distribution of the software without prior 
 * written permission from OSF.  OSF makes no representations about the 
 * suitability of this software for any purpose.  It is provided "as is" 
 * without express or implied warranty. 
 */
/*
 * HISTORY
 * $Log: def_build.c,v $
 * Revision 1.3.4.5  1993/04/28  14:35:49  damon
 * 	CR 463. More pedantic changes
 * 	[1993/04/28  14:34:40  damon]
 *
 * Revision 1.3.4.4  1993/04/27  20:10:07  damon
 * 	CR 463. Made pedantic changes
 * 	[1993/04/27  20:09:56  damon]
 * 
 * Revision 1.3.4.3  1993/04/08  16:30:02  damon
 * 	CR 446. Clean up include files
 * 	[1993/04/08  16:28:07  damon]
 * 
 * Revision 1.3.4.2  1993/01/28  23:38:32  damon
 * 	CR 417. Removed set parameter from default_build
 * 	[1993/01/28  23:38:17  damon]
 * 
 * Revision 1.3.2.6  1992/12/03  17:20:36  damon
 * 	ODE 2.2 CR 183. Added CMU notice
 * 	[1992/12/03  17:07:59  damon]
 * 
 * Revision 1.3.2.5  1992/09/24  19:01:26  gm
 * 	CR282: Made more portable to non-BSD systems.
 * 	[1992/09/23  18:21:32  gm]
 * 
 * Revision 1.3.2.4  1992/08/07  15:38:45  damon
 * 	CR 266. Changed salloc to strdup
 * 	[1992/08/07  15:37:07  damon]
 * 
 * Revision 1.3.2.3  1992/07/26  17:30:34  gm
 * 	Fixed to remove warnings when compiling under OSF/1 R1.1.
 * 	[1992/07/14  17:16:25  gm]
 * 
 * Revision 1.3.2.2  1992/02/18  22:04:23  damon
 * 	Changes for LBE removal
 * 	[1992/02/18  21:57:05  damon]
 * 
 * Revision 1.3  1991/12/05  21:04:31  devrcs
 * 	Moved get_rc_value from def_build to par_rc_file
 * 	[91/04/12  15:46:58  damon]
 * 
 * 	Changed sdm to ode; std_defs.h to  odedefs.h
 * 	[91/01/10  11:45:46  randyb]
 * 
 * 	Changed functions to use pointers instead of arrays.
 * 	Made rm_newline a public utility.
 * 	[91/01/08  12:09:49  randyb]
 * 
 * 	Added include of <sys/param.h>
 * 	[90/12/12  09:23:22  mckeen]
 * 
 * 	Changed the include of sys/access.h to be sys/file.h
 * 	[90/12/11  11:43:08  mckeen]
 * 
 * 	Took out an unnecessary variable from get_rc_value, the sb rc file
 * 	Can now call build_base_dir and get the default build_list; this
 * 	was done by making a new Boolean variable
 * 	[90/12/10  13:27:16  randyb]
 * 
 * 	Added ability to determine backing tree by using the rc variable
 * 	[90/12/07  13:56:32  randyb]
 * 
 * 	Added routine build_base_dir to determine the base directory of builds.
 * 	[90/12/06  11:49:58  randyb]
 * 
 * 	Fixed improper test in get_rc_value to prevent core dump
 * 	[90/12/06  09:40:47  damon]
 * 
 * 	Added argument to get_rc_value to conditionally supress
 * 	"line missing" error message.
 * 	[90/12/06  09:22:06  damon]
 * 
 * 	Pre-OSF1.0 changes
 * 	[90/11/13  17:02:35  randyb]
 * 
 * $EndLog$
 */
/******************************************************************************
**                          Open Software Foundation              	     **
**                              Cambridge, MA                                **
**                              Randy Barbano                                **
**                               April 1990                                  **
*******************************************************************************
**
**  Description:
**	These are functions for library libsb.a.
**
**  Lib Functions and Their Usage:
**	1) default_build ( char** base, char** build, char** set,
**			   char* sb_rcfile) int
**
**	   returns: ERROR if any type of failure else it returns OK.
**
**	   usage:
**	     The values of the first three pointers are filled with their
**	     values from the rc files, sb_rcfile.  If the pointer = NULL,
**	     there no attempt to fill it.
**
**	2) build_base_dir ( char* keyword, char* sbfile,
**			    struct rcfile contents, BOOLEAN errors,
**			    BOOLEAN default ) char *
**
**	   returns: pointer to string or NULL if none found
**
**	    usage:
**	      The value of the keyword is searched for in the build list
**	      file.  This file is found using the rc value of BUILD_LIST.
**	      keyword should be a build name and the return value will be
**	      the base directory to that build.
**
**    functions called by lib functions:
**	1) default_build: parse_rc_file, get_rc_value
**	3) build_base_dir: get_rc_value, nxtarg
**
 */

#include <stdio.h>
#include <string.h>
#include <ode/interface.h>
#include <ode/odedefs.h>
#include <ode/parse_rc_file.h>
#include <ode/util.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <unistd.h>

int	default_build ( 
    char     ** base,                            /* base of project's builds */
    char     ** build,                    /* name of project's default build */
    char      * sb_rcfile )                      /* sandbox rc file to parse */

	/* This function retreives the values for each of the arguments:
	   base, build, and set, assuming they are not set to to NULL.
	   The values come from reading the sandbox rc file.  If all
	   goes well, OK is return, else ERROR. */


{
    struct      rcfile    rc_contents;          /* holds contents of rc file */
    int         value = OK;                                  /* return value */

  memset ( (char *)&rc_contents, 0, sizeof ( rc_contents ));

  if ( parse_rc_file ( sb_rcfile, &rc_contents )) {
    ui_print ( VWARN, "unable to parse sandbox rcfile %s.\n", sb_rcfile );
    value = ERROR;
  } /* if */

  if ( value == OK && build != NULL )
    value = get_rc_value ( DEF_BUILD, build, &rc_contents, TRUE );
  
  if ( value == OK && base != NULL )
    value = get_rc_value ( BUILD_BASE, base, &rc_contents, TRUE );
  
  return ( value );
}                                                           /* default build */

char *
build_base_dir ( 
    char      * keyword,       /* identifying build's record in "build_list" */
    struct      rcfile  * contents,             /* holds contents of rc file */
    BOOLEAN     errors,            /* true if error messages should be given */
    BOOLEAN     defaultx ) 			   /* use default build_list */

	/* This function looks for the base directory of the keyword;
	   assuming the keyword is a build.  It looks for it in the
	   file designated by the rc variable, BUILD_LIST, or in the
	   file, DEFAULT_BLIST.  It returns the base directory or NUL
	   if none is found. */


{
    FILE      * bptr;                              /* ptr to file build_list */
    struct      stat statb;                                /* stat structure */
    char        recbuf [ PATH_LEN ],      /* read of records from build_list */
		listbuf [ PATH_LEN ],   /* hold space for name of build_list */
              * blist,           	      /* path and name of build_list */
	      * rec_ptr,                           /* ptr to recbuf contents */
              * token;                         /* holds each token as parsed */
    const char * errormsg = "ERROR: build_base_dir:";   /* head for error msg */

  if ( *keyword == NUL ) {
    if ( errors )
      ui_print( VWARN, "%s bad keyword argument.\n", errormsg );
    return ( NULL );
  } /* if */

  if ( defaultx ) {        				      /* use default */
    strcpy ( listbuf, DEFAULT_BLIST );
    blist = listbuf;
  } /* if */

  else if ( get_rc_value ( BUILD_LIST, &blist, contents, errors ) == ERROR ) {
    strcpy ( listbuf, DEFAULT_BLIST );  /* if build_list not given, try this */
    blist = listbuf;
  } /* else if */

  if (( access ( blist, R_OK ) != OK ) ||
      (( bptr = fopen ( blist, READ )) == NULL )) {
    if ( streq (  blist, DEFAULT_BLIST )) {
      if ( errors )  			   /* error if already using default */
	ui_print ( VWARN, "%s cannot read build list: %s.\n", errormsg, blist );
      return( NULL );
    } /* if */
 							      /* try default */
    else if (( access ( DEFAULT_BLIST, R_OK ) != OK ) ||
             (( bptr = fopen ( DEFAULT_BLIST, READ )) == NULL )) { 
      if ( errors )
	ui_print ( VWARN, "%s cannot read default build list: %s.\n",
			   errormsg, DEFAULT_BLIST );
      return( NULL );
    } /* else if */
  } /* if */

  while ( fgets ( recbuf, PATH_LEN, bptr) != NULL ) {
    rec_ptr = recbuf;
    token = nxtarg ( &rec_ptr, WHITESPACE );
	  
    if ( *token != NUL && streq ( token, keyword )) {
      token = nxtarg ( &rec_ptr, WHITESPACE );            /* get third token */

      if ( *token == NUL || *rec_ptr == NUL ) {
        if ( errors )
	  ui_print ( VWARN, "%s no base directory for build %s.\n",
			     errormsg, keyword );
        return( NULL );
      } /* if */

      rm_newline ( rec_ptr );

      if (( stat ( rec_ptr, &statb ) == ERROR ) ||
	  (( statb.st_mode & S_IFMT ) != S_IFDIR )) {
        if ( errors )
	  ui_print ( VWARN, "%s base for build %s is not a directory.\n",
			     errormsg, keyword );
        return( NULL );
      } /* if */

      if (( rec_ptr = (char *) strdup ( rec_ptr )) == NULL ) {
	ui_print ( VWARN, "%s unable to strdup base dir.\n", errormsg );
        return( NULL );
      } /* if */
	
      return( rec_ptr );
    } /* if */
  } /* while */

  if ( errors )
    ui_print ( VWARN, "%s build %s not found in build list: %s.\n", errormsg,
		      keyword, blist );

  return ( NULL );
}                                                          /* build base dir */

